azuredevops/auth.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2023 The Flux CD contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azuredevops
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/fluxcd/go-git-providers/gitprovider"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/v6"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/v6/core"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/v6/git"
+
+	"net/url"
+)
+
+// NewClient creates a new Client instance for Azure Devops API endpoints.
+// The client accepts a personal token used which is used to authenticate and a context as an argument,
+// Variadic parameters gitprovider.ClientOption are used to pass additional options to the gitprovider.Client.
+func NewClient(personalAccessToken string, ctx context.Context, optFns ...gitprovider.ClientOption) (gitprovider.Client, error) {
+	opts, err := gitprovider.MakeClientOptions(optFns...)
+	if err != nil {
+		return nil, err
```

Comment:
We should entertain the habit of wrapping errors and add context:
```suggestion
		return nil, fmt.Errorf("failed making client options: %w", err)
```